File: sdk/src/turns.rs
//! # The authorized turn builder — the SDK's one public turn shape.
//!
//! ```text
//! Identity → .turn() → typed verb builders → .sign() → .submit() → Receipt
//! ```
//!
//! [`AgentRuntime::turn()`](crate::AgentRuntime::turn) opens a
//! [`TurnBuilder`]; the typed verbs ([`transfer`](TurnBuilder::transfer),
//! [`write`](TurnBuilder::write), [`grant`](TurnBuilder::grant), …, or a
//! [`factories`](crate::factories)/[`polis`](crate::polis) plan via
//! [`effects`](TurnBuilder::effects)) accumulate the act;
//! [`sign()`](TurnBuilder::sign) binds it to the identity's Ed25519 key over
//! the canonical signing message (federation-bound, replay-separated); and
//! [`submit()`](AuthorizedTurn::submit) executes it and returns the
//! [`Receipt`] noun.
//!
//! **An unauthorized act is inexpressible here.** There is no method on this
//! surface that yields an unsigned action — by the time anything reaches the
//! executor it carries a real `Authorization::Signature`. The raw vocabulary
//! (including the genesis-only `Authorization::Unchecked`) lives behind the
//! sealed [`raw`](crate::raw) module.
//!
//! The anti-blind-signing affordance rides along:
//! [`AuthorizedTurn::explain()`] renders the clerk's faithful, total
//! explanation of exactly what was signed.

use dregg_cell::state::FieldElement;
use dregg_cell::{CapabilityRef, CellId, field_from_u64};
use dregg_turn::Effect;

use crate::error::SdkError;
use crate::raw;
use crate::receipt::Receipt;
use crate::runtime::AgentRuntime;

/// Who acts and pays for the turn being built.
#[derive(Clone, Copy, Debug)]
enum Acting {
    /// Ordinary agent turn: the runtime's own agent cell acts and pays.
    Agent,
    /// Agent-paid turn whose ACTION targets another cell the identity
    /// administers (signature verified against the target's `owner_pubkey`;
    /// parent-gate capability required). The production shape for driving
    /// factory-born cells.
    On(CellId),
    /// Cell-agent turn: `cell` is the turn agent AND action target, paying
    /// `fee` from its own balance (the one-time adopt bootstrap shape).
    AsCell(CellId, u64),
}

/// The typed verb builder. Open one with
/// [`AgentRuntime::turn()`](crate::AgentRuntime::turn); finish with
/// [`sign()`](Self::sign).
#[derive(Debug)]
pub struct TurnBuilder<'rt> {
    runtime: &'rt AgentRuntime,
    acting: Acting,
    method: String,
    effects: Vec<Effect>,
    witness_blobs: Vec<dregg_turn::action::WitnessBlob>,
    fee: Option<u64>,
}

impl<'rt> TurnBuilder<'rt> {
    pub(crate) fn new(runtime: &'rt AgentRuntime) -> Self {
        TurnBuilder {
            runtime,
            acting: Acting::Agent,
            method: "execute".to_string(),
            effects: Vec::new(),
            witness_blobs: Vec::new(),
            fee: None,
        }
    }

    /// The cell whose authority this turn exercises (the default `from` /
    /// write target for the typed verbs).
    fn acting_cell(&self) -> CellId {
        match self.acting {
            Acting::Agent => self.runtime.cell_id(),
            Acting::On(t) => t,
            Acting::AsCell(c, _) => c,
        }
    }

    /// Target another cell the identity administers (the action targets
    /// `target`; this agent signs and pays). The executor verifies the
    /// signature against `target`'s `owner_pubkey` and requires the agent's
    /// c-list capability on it — the [`AgentRuntime::execute_on`] shape.
    pub fn on(mut self, target: CellId) -> Self {
        self.acting = Acting::On(target);
        self
    }

    /// Act AS `cell` (cell-agent turn): `cell` is the turn agent and pays
    /// `fee` from its own balance — the [`AgentRuntime::execute_as`] shape
    /// used for the one-time factory adopt bootstrap.
    pub fn as_cell(mut self, cell: CellId, fee: u64) -> Self {
        self.acting = Acting::AsCell(cell, fee);
        self
    }

    /// Set the action's method verb (default `"execute"`). Workers under a
    /// scoped capability credential are admitted per-method.
    pub fn method(mut self, name: &str) -> Self {
        self.method = name.to_string();
        self
    }

    /// Set the turn fee (computron budget). Defaults to the runtime's
    /// standard fee (or the `as_cell` fee).
    pub fn fee(mut self, fee: u64) -> Self {
        self.fee = Some(fee);
        self
    }

    // ─── typed verbs ───

    /// Transfer `amount` computrons from the acting cell to `to`.
    pub fn transfer(mut self, to: CellId, amount: u64) -> Self {
        let from = self.acting_cell();
        self.effects.push(Effect::Transfer { from, to, amount });
        self
    }

    /// Transfer with an explicit source cell (must still be within this
    /// identity's authority — the executor checks, not the builder).
    pub fn transfer_from(mut self, from: CellId, to: CellId, amount: u64) -> Self {
        self.effects.push(Effect::Transfer { from, to, amount });
        self
    }

    /// Write state slot `index` of the acting cell (the `write` verb;
    /// admitted only where the cell's installed program allows).
    pub fn write(mut self, index: usize, value: FieldElement) -> Self {
        let cell = self.acting_cell();
        self.effects.push(Effect::SetField { cell, index, value });
        self
    }

    /// [`write`](Self::write) with a numeric value (encoded like
    /// [`field_from_u64`]).
    pub fn write_u64(self, index: usize, value: u64) -> Self {
        self.write(index, field_from_u64(value))
    }

    /// Grant a capability from the acting cell to `to` (the `grant` verb —
    /// non-amplifying: the executor admits only grants within held
    /// authority).
    pub fn grant(mut self, to: CellId, cap: CapabilityRef) -> Self {
        let from = self.acting_cell();
        self.effects.push(Effect::GrantCapability { from, to, cap });
        self
    }

    /// Bump the acting cell's nonce (a deliberate no-op state advance).
    pub fn increment_nonce(mut self) -> Self {
        let cell = self.acting_cell();
        self.effects.push(Effect::IncrementNonce { cell });
        self
    }

    /// Append one prebuilt effect (escape for verbs without dedicated
    /// sugar; the executor's gates apply identically).
    pub fn effect(mut self, effect: Effect) -> Self {
        self.effects.push(effect);
        self
    }

    /// Append a prebuilt effect list — the splice point for the
    /// [`factories`](crate::factories) / [`polis`](crate::polis) /
    /// [`program`](crate::program) plan builders (`plan.create_effects`,
    /// `release_escrow(..)`, `propose(..)`, …).
    pub fn effects(mut self, effects: impl IntoIterator<Item = Effect>) -> Self {
        self.effects.extend(effects);
        self
    }

    /// Exhibit a 32-byte preimage witness with this turn (the `reveal`
    /// verb). The blob rides `Action::witness_blobs` UNDER the signature
    /// and is what `PreimageGate` / `KeyRotationGate` cell programs verify
    /// against the committed digest — the identity pre-rotation rotate
    /// turn carries the presented key-set commitment this way.
    pub fn reveal(mut self, preimage: [u8; 32]) -> Self {
        self.witness_blobs.push(dregg_turn::action::WitnessBlob {
            kind: dregg_turn::action::WitnessKind::Preimage32,
            bytes: preimage.to_vec(),
        });
        self
    }

    // ─── terminal ───

    /// Sign the built action with this identity's key over the canonical
    /// federation-bound signing message, yielding an [`AuthorizedTurn`]
    /// ready to [`submit`](AuthorizedTurn::submit).
    ///
    /// After this point the act is credentialed; there is no way back to an
    /// unauthorized shape.
    pub fn sign(self) -> Result<AuthorizedTurn<'rt>, SdkError> {
        if self.effects.is_empty() {
            return Err(SdkError::Rejected(
                "refusing to sign an empty turn (no effects staged)".to_string(),
            ));
        }
        let target = self.acting_cell();
        let mut unsigned = raw::unsigned_action_named(target, &self.method, self.effects);
        // Witnesses are attached BEFORE signing so the signature covers
        // them (the `set_field_with_preimage` shape in the executor's
        // coverage tests).
        unsigned.witness_blobs = self.witness_blobs;
        let action = self.runtime.sign_action_for_runtime(unsigned);
        Ok(AuthorizedTurn {
            runtime: self.runtime,
            acting: self.acting,
            action,
            fee: self.fee,
        })
    }
}

/// A signed, ready-to-submit turn. Produced by [`TurnBuilder::sign`];
/// consumed by [`submit`](Self::submit).
#[derive(Debug)]
pub struct AuthorizedTurn<'rt> {
    runtime: &'rt AgentRuntime,
    acting: Acting,
    action: dregg_turn::Action,
    fee: Option<u64>,
}

impl AuthorizedTurn<'_> {
    /// The clerk's faithful, total explanation of exactly what was signed
    /// (the anti-blind-signing reading; see [`crate::explain`]).
    pub fn explain(&self) -> String {
        crate::explain::explain_action(&self.action)
    }

    /// The signed action (inspection only — `submit` consumes `self`).
    pub fn action(&self) -> &dregg_turn::Action {
        &self.action
    }

    /// Execute the turn and return the [`Receipt`] noun.
    ///
    /// Routing follows the builder's acting mode: an ordinary agent turn
    /// (or `.on(target)`) is agent-paid and appended to the identity's
    /// receipt chain; an `.as_cell(..)` turn is paid by the cell and
    /// belongs to the cell's history.
    pub fn submit(self) -> Result<Receipt, SdkError> {
        let receipt = match self.acting {
            Acting::Agent | Acting::On(_) => self
                .runtime
                .submit_signed_action_as_agent(self.action, self.fee.unwrap_or(10_000))?,
            Acting::AsCell(cell, cell_fee) => self.runtime.submit_signed_action_as_cell(
                cell,
                self.action,
                self.fee.unwrap_or(cell_fee),
            )?,
        };
        Ok(Receipt::new(receipt))
    }
}

File: sdk-ts/src/turns.ts
/**
 * The authorized turn builder — the SDK's one public turn shape.
 *
 * ```text
 * Identity → .turn() → typed verb builders → .sign() → .submit() → Receipt
 * ```
 *
 * `AgentRuntime.turn()` opens a [`TurnBuilder`]; the typed verbs
 * ([`transfer`], [`write`], [`grant`], …) accumulate the act; [`sign`] binds
 * it to the identity's Ed25519 key over the canonical signing message
 * (federation-bound, replay-separated); and [`AuthorizedTurn.submit`]
 * executes it on the node and returns the [`Receipt`] noun.
 *
 * **An unauthorized act is inexpressible here.** No method on this surface
 * yields an unsigned action — by the time anything reaches the node it
 * carries a real `Authorization::Signature` per action AND a `SignedTurn`
 * envelope signature over the canonical turn hash. The raw vocabulary
 * (including unauthorized construction) lives behind the sealed
 * `@dregg/sdk/raw` module.
 *
 * The anti-blind-signing affordance rides along: [`AuthorizedTurn.explain`]
 * renders the clerk's faithful, total explanation of exactly what was
 * signed.
 *
 * Difference from the Rust builder: there is no `.as_cell(..)` here — the
 * remote signed ingress pins `turn.agent` to the signer's default cell
 * (node/src/api.rs `post_submit_signed_turn`), so a cell-agent turn cannot
 * be expressed over this transport. `.on(target)` (act on another cell the
 * identity administers, signature verified against the target's
 * `owner_pubkey`) works as in Rust.
 */

import type { AgentRuntime } from "./client";
import { Receipt } from "./receipt";
import { explainAction } from "./explain";
import type { Action, Bytes32, CapabilityRef, CellId, Effect, Turn } from "./internal/wire";
import { fieldFromU64, unsignedActionNamed } from "./internal/wire";

/** The canonical `Payable` `pay` method name (mirrors `dregg_payable::PAY_METHOD`). */
export const PAY_METHOD = "pay";

/** Default computron budget when `.fee()` is not called (Rust parity). */
const DEFAULT_FEE = 10_000n;

/** Turn validity horizon stamped on submitted turns (seconds). */
const VALIDITY_HORIZON_SECS = 3600n;

/** Refusal to sign a meaningless turn, mirroring the Rust builder. */
export class EmptyTurnError extends Error {
  constructor() {
    super("refusing to sign an empty turn (no effects staged)");
    this.name = "EmptyTurnError";
  }
}

/**
 * The typed verb builder. Open one with `runtime.turn()`; finish with
 * [`sign`].
 */
export class TurnBuilder {
  private readonly runtime: AgentRuntime;
  /** `undefined` = ordinary agent turn; a CellId = the `.on(target)` shape. */
  private actingOn: CellId | undefined;
  private methodName = "execute";
  private effectList: Effect[] = [];
  private argList: Bytes32[] = [];
  private feeValue: bigint | undefined;

  constructor(runtime: AgentRuntime) {
    this.runtime = runtime;
  }

  /** The cell whose authority this turn exercises. */
  private actingCell(): CellId {
    return this.actingOn ?? this.runtime.identity.cellId();
  }

  /**
   * Target another cell the identity administers (the action targets
   * `target`; this agent signs and pays). The node verifies the signature
   * against `target`'s `owner_pubkey` and requires the agent's c-list
   * capability on it.
   */
  on(target: CellId): this {
    this.actingOn = target;
    return this;
  }

  /** Set the action's method verb (default `"execute"`). */
  method(name: string): this {
    this.methodName = name;
    return this;
  }

  /** Set the turn fee (computron budget). Defaults to 10 000. */
  fee(fee: number | bigint): this {
    this.feeValue = BigInt(fee);
    return this;
  }

  // ─── typed verbs ───

  /** Transfer `amount` computrons from the acting cell to `to`. */
  transfer(to: CellId, amount: number | bigint): this {
    this.effectList.push({ kind: "transfer", from: this.actingCell(), to, amount });
    return this;
  }

  /**
   * Transfer with an explicit source cell (must still be within this
   * identity's authority — the executor checks, not the builder).
   */
  transferFrom(from: CellId, to: CellId, amount: number | bigint): this {
    this.effectList.push({ kind: "transfer", from, to, amount });
    return this;
  }

  /**
   * Write state slot `index` of the acting cell (admitted only where the
   * cell's installed program allows).
   */
  write(index: number, value: Bytes32): this {
    this.effectList.push({ kind: "setField", cell: this.actingCell(), index, value });
    return this;
  }

  /** [`write`] with a numeric value (encoded like `field_from_u64`). */
  writeU64(index: number, value: number | bigint): this {
    return this.write(index, fieldFromU64(value));
  }

  /**
   * Grant a capability from the acting cell to `to` (non-amplifying: the
   * executor admits only grants within held authority).
   */
  grant(to: CellId, cap: CapabilityRef): this {
    this.effectList.push({ kind: "grantCapability", from: this.actingCell(), to, cap });
    return this;
  }

  /** Bump the acting cell's nonce (a deliberate no-op state advance). */
  incrementNonce(): this {
    this.effectList.push({ kind: "incrementNonce", cell: this.actingCell() });
    return this;
  }

  /** Append one prebuilt effect (escape hatch; the executor's gates apply identically). */
  effect(effect: Effect): this {
    this.effectList.push(effect);
    return this;
  }

  /** Append a prebuilt effect list (the splice point for plan builders). */
  effects(effects: Iterable<Effect>): this {
    for (const e of effects) this.effectList.push(e);
    return this;
  }

  /**
   * Set the action's argument vector (the typed witness the method carries;
   * the routing/auth gate on the method symbol, these are the receipt-bound
   * record). Each entry is a 32-byte field element. Replaces any prior args.
   */
  args(args: Bytes32[]): this {
    this.argList = args.slice();
    return this;
  }

  /**
   * **`pay`** — move `amount` of `asset` from the acting cell to `to` through
   * the canonical `Payable` `pay` desugar. The byte-identical twin of
   * `dregg_payable::resolve_pay` / the Rust SDK's `AgentRuntime::pay`: the
   * action's `method` is `pay`, its `args` are `[asset, field_from_u64(amount),
   * to]` (the `pay_args` witness), and it carries EXACTLY ONE conserving
   * `Effect::Transfer` (per-asset Σδ=0). The same value rail the app
   * framework's `Payable::pay` and the metered tool-gateway charge ride — not a
   * hand-rolled effect.
   *
   * `asset` is the asset to pay in (the payer's `token_id`; a bridged `$DREGG`
   * mirror asset is an ordinary 32-byte id, routed identically).
   */
  pay(to: CellId, amount: number | bigint, asset: Bytes32): this {
    this.methodName = PAY_METHOD;
    this.argList = [asset, fieldFromU64(amount), to];
    this.effectList.push({ kind: "transfer", from: this.actingCell(), to, amount });
    return this;
  }

  // ─── terminal ───

  /**
   * Sign the built action with this identity's key over the canonical
   * federation-bound signing message, yielding an [`AuthorizedTurn`] ready
   * to [`submit`](AuthorizedTurn.submit).
   *
   * After this point the act is credentialed; there is no way back to an
   * unauthorized shape. (Async because the federation binding is discovered
   * from the node on first use.)
   */
  async sign(): Promise<AuthorizedTurn> {
    if (this.effectList.length === 0) {
      throw new EmptyTurnError();
    }
    const target = this.actingCell();
    const federationId = await this.runtime.node.federationId();
    const unsigned = unsignedActionNamed(target, this.methodName, this.effectList);
    unsigned.args = this.argList;
    const action = this.runtime.identity.signAction(unsigned, federationId);
    return new AuthorizedTurn(this.runtime, action, this.feeValue ?? DEFAULT_FEE);
  }
}

/**
 * A signed, ready-to-submit turn. Produced by [`TurnBuilder.sign`]; consumed
 * by [`submit`](AuthorizedTurn.submit).
 */
export class AuthorizedTurn {
  private readonly runtime: AgentRuntime;
  private readonly signedAction: Action;
  private readonly fee: bigint;
  private submitted = false;

  constructor(runtime: AgentRuntime, action: Action, fee: bigint) {
    this.runtime = runtime;
    this.signedAction = action;
    this.fee = fee;
  }

  /**
   * The clerk's faithful, total explanation of exactly what was signed —
   * the anti-blind-signing reading (see `explain.ts`).
   */
  explain(): string {
    return explainAction(this.signedAction);
  }

  /** The signed action (inspection only — `submit` consumes the turn). */
  action(): Action {
    return this.signedAction;
  }

  /**
   * Execute the turn on the node and return the [`Receipt`] noun.
   *
   * The agent cell pays; the turn rides the cell's live nonce, the node's
   * receipt-chain head (`previous_receipt_hash` causal binding), and a
   * one-hour validity horizon; the envelope signature binds the canonical
   * `Turn::hash` (v3). A chain-head race (another commit landing between
   * read and submit) is retried once with fresh bindings — the per-action
   * signature stays valid; only the envelope is re-signed. One-shot: a
   * second call is refused (the consumed turn would replay-fail anyway).
   */
  async submit(): Promise<Receipt> {
    if (this.submitted) {
      throw new Error("AuthorizedTurn already submitted (one-shot, like the Rust consume-on-submit)");
    }
    this.submitted = true;
    let lastError: unknown;
    for (let attempt = 0; attempt < 2; attempt++) {
      const nonce = await this.runtime.currentNonce();
      const previousReceiptHash = await this.runtime.node.receiptChainHead();
      const turn: Turn = {
        agent: this.runtime.identity.cellId(),
        nonce,
        roots: [{ action: this.signedAction, children: [] }],
        fee: this.fee,
        validUntil: BigInt(Math.floor(Date.now() / 1000)) + VALIDITY_HORIZON_SECS,
        previousReceiptHash,
      };
      try {
        return await this.runtime.submitTurn(turn);
      } catch (e) {
        lastError = e;
        const msg = e instanceof Error ? e.message : String(e);
        if (attempt === 0 && /receipt chain mismatch|nonce/i.test(msg)) {
          continue; // racing commit moved the head; rebind and retry once
        }
        throw e;
      }
    }
    throw lastError;
  }
}
